src/c2s/mongoose_c2s.erl

```diff
@@ -0,0 +1,920 @@
+    {stop, {shutdown, retries}};
+maybe_retry_state({wait_for_sasl_response, _, 0}) ->
+    {stop, {shutdown, retries}};
+maybe_retry_state({wait_for_feature_after_auth, Retries}) ->
+    {wait_for_feature_after_auth, Retries - 1};
+maybe_retry_state({wait_for_feature_before_auth, SaslState, Retries}) ->
+    {wait_for_feature_before_auth, SaslState, Retries - 1};
+maybe_retry_state({wait_for_sasl_response, SaslState, Retries}) ->
+    {wait_for_sasl_response, SaslState, Retries - 1}.
+
+%% @doc Check 'from' attribute in stanza RFC 6120 Section 8.1.2.1
+-spec verify_from(exml:element(), jid:jid()) -> boolean().
+verify_from(El, StateJid) ->
+    case exml_query:attr(El, <<"from">>) of
+        undefined -> true;
+        SJid ->
+            jid:are_equal(jid:from_binary(SJid), StateJid)
+    end.
+
+-spec handle_foreign_packet(c2s_data(), c2s_state(), exml:element()) -> fsm_res().
+handle_foreign_packet(StateData = #c2s_data{host_type = HostType, lserver = LServer}, C2SState, El) ->
+    ?LOG_DEBUG(#{what => packet_before_session_established_sent, packet => El, c2s_pid => self()}),
+    ServerJid = jid:make_noprep(<<>>, LServer, <<>>),
+    AccParams = #{host_type => HostType, lserver => LServer, location => ?LOCATION,
+                  element => El, from_jid => ServerJid, to_jid => ServerJid},
+    Acc0 = mongoose_acc:new(AccParams),
+    Params = hook_arg(StateData, C2SState),
+    {_, Acc1} = mongoose_c2s_hooks:user_send_xmlel(HostType, Acc0, Params),
+    handle_state_after_packet(StateData, C2SState, Acc1).
+
+-spec handle_c2s_packet(c2s_data(), c2s_state(), exml:element()) -> fsm_res().
+handle_c2s_packet(StateData = #c2s_data{host_type = HostType}, C2SState, El) ->
+    Acc0 = element_to_origin_accum(StateData, El),
+    Acc1 = mongoose_hooks:c2s_preprocessing_hook(HostType, Acc0, StateData),
+    case mongoose_acc:get(hook, result, undefined, Acc1) of
+        drop -> {next_state, session_established, StateData};
+        _ -> do_handle_c2s_packet(StateData, C2SState, Acc1)
+    end.
+
+-spec do_handle_c2s_packet(c2s_data(), c2s_state(), mongoose_acc:t()) -> fsm_res().
+do_handle_c2s_packet(StateData = #c2s_data{host_type = HostType}, C2SState, Acc) ->
+    case mongoose_c2s_hooks:user_send_packet(HostType, Acc, hook_arg(StateData)) of
+        {ok, Acc1} ->
+            Acc2 = handle_outgoing_stanza(StateData, Acc1, mongoose_acc:stanza_name(Acc1)),
+            handle_state_after_packet(StateData, C2SState, Acc2);
+        {stop, Acc1} ->
+            handle_state_after_packet(StateData, C2SState, Acc1)
+    end.
+
+%% @doc Process packets sent by the user (coming from user on c2s XMPP connection)
+-spec handle_outgoing_stanza(c2s_data(), mongoose_acc:t(), binary()) -> mongoose_acc:t().
+handle_outgoing_stanza(StateData = #c2s_data{host_type = HostType}, Acc, <<"message">>) ->
+    TS0 = mongoose_acc:timestamp(Acc),
+    Acc1 = mongoose_c2s_hooks:user_send_message(HostType, Acc, hook_arg(StateData)),
```

Comment:
As in the answer for `user_receive_packet`.

Reply:
Also keeping "d" instead of "t" because the old hook was already named `user_send_packet` and thought of keeping it backwards compatible. Changing the name of the hook means having to also update MAM and inbox and all others and this PR is sufficiently enormous 😅